core/types/l2trace.go

```diff
@@ -0,0 +1,165 @@
+package types
+
+import (
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/scroll-tech/go-ethereum/rlp"
+)
+
+// BlockResult contains block execution traces and results required for rollers.
+type BlockResult struct {
+	ExecutionResults []*ExecutionResult `json:"executionResults"`
+}
+
+type rlpBlockResult struct {
+	ExecutionResults []*ExecutionResult
+}
+
+func (b *BlockResult) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, &rlpBlockResult{
+		ExecutionResults: b.ExecutionResults,
+	})
+}
+
+func (b *BlockResult) DecodeRLP(s *rlp.Stream) error {
+	var dec rlpBlockResult
+	err := s.Decode(&dec)
+	if err == nil {
+		b.ExecutionResults = dec.ExecutionResults
+	}
+	return err
+}
+
+// ExecutionResult groups all structured logs emitted by the EVM
+// while replaying a transaction in debug mode as well as transaction
+// execution status, the amount of gas used and the return value
+type ExecutionResult struct {
+	Gas         uint64         `json:"gas"`
+	Failed      bool           `json:"failed"`
+	ReturnValue string         `json:"returnValue,omitempty"`
+	StructLogs  []StructLogRes `json:"structLogs"`
+}
+
+type rlpExecutionResult struct {
+	Gas         uint64
+	Failed      bool
+	ReturnValue string
+	StructLogs  []StructLogRes
+}
+
+func (e *ExecutionResult) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, rlpExecutionResult{
+		Gas:         e.Gas,
+		Failed:      e.Failed,
+		ReturnValue: e.ReturnValue,
+		StructLogs:  e.StructLogs,
+	})
+}
+
+func (e *ExecutionResult) DecodeRLP(s *rlp.Stream) error {
+	var dec rlpExecutionResult
+	err := s.Decode(&dec)
+	if err == nil {
+		e.Gas, e.Failed, e.ReturnValue, e.StructLogs = dec.Gas, dec.Failed, dec.ReturnValue, dec.StructLogs
+	}
+	return err
+}
+
+// StructLogRes stores a structured log emitted by the EVM while replaying a
+// transaction in debug mode
+type StructLogRes struct {
+	Pc      uint64             `json:"pc"`
+	Op      string             `json:"op"`
+	Gas     uint64             `json:"gas"`
+	GasCost uint64             `json:"gasCost"`
+	Depth   int                `json:"depth"`
+	Error   string             `json:"error,omitempty"`
+	Stack   *[]string          `json:"stack,omitempty"`
+	Memory  *[]string          `json:"memory,omitempty"`
+	Storage *map[string]string `json:"storage,omitempty"`
+}
+
+type rlpStructLogRes struct {
+	Pc      uint64   `json:"pc"`
```

Comment:
also remove the json format for this rlp struct?

Reply:
removed.